Week4/week 4.ipynb

```diff
@@ -0,0 +1,335 @@
+    "\n",
+    "    curobj.execute(log_tbl_insert_trigger)\n",
+    "    log_table_delete_trigger=\"create trigger IF NOT EXISTS delete_trigger\\\n",
+    "                                after delete \\\n",
+    "                                on contact\\\n",
+    "                                begin\\\n",
+    "                                insert into logtable\\\n",
+    "                                values(old.fname,old.lname,old.contact,datetime('now','localtime'),'DELETE');\\\n",
+    "                                end;\"\n",
+    "\n",
+    "    curobj.execute(log_table_delete_trigger)\n",
+    "    log_table_update_trigger=\"create trigger IF NOT EXISTS update_trigger\\\n",
+    "                                after update\\\n",
+    "                                on contact\\\n",
+    "                                begin\\\n",
+    "                                insert into logtable\\\n",
+    "                                values(new.fname,new.lname,new.contact,datetime('now','localtime'),'UPDATE(After)');\\\n",
+    "                                insert into logtable\\\n",
+    "                                values(old.fname,old.lname,old.contact,datetime('now','localtime'),'UPDATE(Before)');\\\n",
+    "                                end;\"\n",
+    "\n",
+    "    curobj.execute(log_table_update_trigger)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "id": "14156ea3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def menu():\n",
+    "    print('Enter 1 To Insert Contact No')\n",
+    "    print('Enter 2 To Delete Contact No')\n",
+    "    print('Enter 3 To Update Contact No')\n",
+    "    print('Enter 4 To Search Contact No')\n",
+    "    print('Enter 0 To Quit')\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "id": "27eae152",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def insert_contact():\n",
+    "    try:\n",
+    "            query = \"insert into contact \\\n",
+    "                    values(?,?,?,?,?)\"\n",
+    "            \n",
+    "            First_Name=input('Enter Your First Name :')\n",
+    "            Last_Name=input('Enter Your Last Name :')\n",
+    "            Contact = int(input('Enter Your Contact No :'))\n",
+    "            Email=input('Enter Your Email ID :')\n",
+    "            City=input('Enter Your City :')\n",
+    "            main_list=[First_Name,Last_Name,Contact,Email,City]\n",
+    "            curobj.execute(query,main_list)\n",
+    "            print()\n",
+    "            print(\"Record(s) are/is Inserted Successfully\")\n",
+    "            print()\n",
+    "    except Error as e:\n",
+    "            print()\n",
+    "            print(e)\n",
+    "            print()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "id": "0f25e53c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def delete_contact():\n",
+    "    try:\n",
```

Comment:
can you explain why try and except are used?

Reply:
When any error occurs when we call program ,
Python will automatically stop the execution of program and give error message.These errors can be handled by (try). So when (try) will raise an error the (except) line will be executed.I will explain it step by step:
1)First, the try clause is executed i.e. the code between try.
2)If there is no exception, then only the try clause will run, except clause is finished.
3)If any exception occurs, the try clause will be skipped and except clause will run.
4)If any exception occurs, but the except clause within the code doesn’t handle it, it is passed on to the outer try statements. If the exception is left unhandled, then the execution stops.